shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/handler/dml/InsertStatementHandler.java

```diff
@@ -0,0 +1,165 @@
+public final class InsertStatementHandler {
+
+    /**
+     * Get OnDuplicateKeyColumnsSegment.
+     *
+     * @param insertStatement InsertStatement
+     * @return OnDuplicateKeyColumnsSegment
+     */
+    public static Optional<OnDuplicateKeyColumnsSegment> getOnDuplicateKeyColumnsSegment(final InsertStatement insertStatement) {
+        if (insertStatement instanceof MySQLStatement) {
+            return ((MySQLInsertStatement) insertStatement).getOnDuplicateKeyColumns();
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Get SetAssignmentSegment.
+     *
+     * @param insertStatement InsertStatement
+     * @return SetAssignmentSegment
+     */
+    public static Optional<SetAssignmentSegment> getSetAssignmentSegment(final InsertStatement insertStatement) {
+        if (insertStatement instanceof MySQLStatement) {
+            return ((MySQLInsertStatement) insertStatement).getSetAssignment();
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Get WithSegment.
+     *
+     * @param insertStatement InsertStatement
+     * @return WithSegment
+     */
+    public static Optional<WithSegment> getWithSegment(final InsertStatement insertStatement) {
+        if (insertStatement instanceof PostgreSQLStatement) {
+            return ((PostgreSQLInsertStatement) insertStatement).getWithSegment();
+        }
+        if (insertStatement instanceof SQLServerStatement) {
+            return ((SQLServerInsertStatement) insertStatement).getWithSegment();
+        }
+        return Optional.empty();
+    }
+    
+    /**
+     * Get column names.
+     *
+     * @param insertStatement InsertStatement
+     * @return column names collection
+     */
+    public static List<String> getColumnNames(final InsertStatement insertStatement) {
+        Optional<SetAssignmentSegment> setAssignment = getSetAssignmentSegment(insertStatement);
+        return setAssignment.isPresent() ? getColumnNamesForSetAssignment(setAssignment.get()) : getColumnNamesForInsertColumns(insertStatement.getColumns());
+    }
+    
+    private static List<String> getColumnNamesForSetAssignment(final SetAssignmentSegment setAssignment) {
+        List<String> result = new LinkedList<>();
+        for (AssignmentSegment each : setAssignment.getAssignments()) {
+            result.add(each.getColumn().getIdentifier().getValue().toLowerCase());
+        }
+        return result;
+    }
+    
+    private static List<String> getColumnNamesForInsertColumns(final Collection<ColumnSegment> columns) {
+        List<String> result = new LinkedList<>();
+        for (ColumnSegment each : columns) {
+            result.add(each.getIdentifier().getValue().toLowerCase());
+        }
+        return result;
+    }
+    
+    /**
+     * Get value count for per value list.
+     *
+     * @param insertStatement InsertStatement
+     * @return value count
+     */
+    public static int getValueCountForPerGroup(final InsertStatement insertStatement) {
```

Comment:
Since `getValueCountForPerGroup` is only used by `GeneratedKeyContextEngine`, do you think it is better to move it to `GeneratedKeyContextEngine`?

Reply:
Here is my thought about this point,
I am a little worried that we mix up the `handler` and `context` for one `SQLStatement`. For me, `handler` is to hide the handling for database dialects and provide the basic functions to get the dialect segments. While, `context` is to give some of the functions containing `ShardingSphere` logics (or simply speaking, for routing and rewriting)
I'd like to listen to your thinking. If I missed something, please let me know. :-)

Reply:
> Since `getValueCountForPerGroup` is only used by `GeneratedKeyContextEngine`, do you think it is better to move it to `GeneratedKeyContextEngine`?

@tristaZero Yes, I agree with you, there is really no other scenario using this method.

Reply:
> Here is my thought about this point,
> I am a little worried that we mix up the `handler` and `context` for one `SQLStatement`. For me, `handler` is to hide the handling for database dialects and provide the basic functions to get the dialect segments. While, `context` is to give some of the functions containing `ShardingSphere` logics (or simply speaking, for routing and rewriting)
> I'd like to listen to your thinking. If I missed something, please let me know. :-)

@tristaZero I agree with your point on the responsibilities for `handler` and `context`, and I will optimize the code according to your suggestions.